RFC - Added the apache_connection_stream() function for CGI WebSockets

RichardTMiles opened the pull request:
The accepted practice for WebSockets (WS) in Apache PHP is to use the `stream_context_create` is a PHP CLI process that Apache’s  `mod_proxy` and `mod_proxy_wstunnel` forward connections to. This requires that you bind to an open port and have access to install Apache modules. The IETF standardized the WebSocket protocol as RFC 6455 in 2011 to enable full duplex communication. The issue Apache2handler PHP currently has is its ability to capture persistent input connections in a php_stream to then be used with stream_select. PHP-defined IO streams are currently not able to handle the requests as `php://stdio` is the actual code buffer and `php://input` is tied to HTTPD's `ap_get_brigade` function which acts as a wrapper to handle io filtering. Not only does `ap_get_brigade` not capture WS input, but it also caches the data as Apache2handler expects it to be static. PHP’s `stream_select` function also fails to cast this non-sealable stream (as expected). HTTPD can handle long upgrade connections, and according to the spec, the WS protocol is just an HTTP GET request with specific headers and io filters. This proposed addition includes a visual sample and example repository.

```PHP
    public static function handleSingleUserConnections(): void
    {

        if (!defined('STDOUT')) {

            define('STDOUT', fopen('php://stdout', 'wb'));

        }

        // get all headers has a polyfill in our function.php
        $headers = getallheaders();

        self::handshake(STDOUT, $headers);

        $flush = self::outputBufferWebSocketEncoder();

        print posix_getpid() . PHP_EOL;

        $flush();

        // Here you can handle the WebSocket upgrade logic
        $websocket = apache_connection_stream();

        if (!is_resource($websocket)) {

            throw new Error('INPUT is not a valid resource');

        }

        $loop = 0;

        while (true) {

            try {

                ++$loop;

                if (!is_resource($websocket)) {

                    throw new Error('STDIN is not a valid resource');

                }

                $flush();

                $read = [$websocket];

                $number = stream_select($read, $write, $error, 10);

                if ($number === 0) {

                    self::colorCode("No streams are requesting to be processed. (loop: $loop; )", 'cyan');

                    continue;

                }

                self::colorCode("$number, stream(s) are requesting to be processed.");

                foreach ($read as $connection) {

                    if ($connection === $websocket) {

                        $data = self::decode($connection);

                        print_r($data);

                        $flush();

                    }

                }

            } catch (Throwable $e) {

                print_r($e);

            }

        }

    }

```


Here is a [link to the repository](https://github.com/RichardTMiles/PHP-Websockets) with the full code example for both PHP-CLI WebSockets and PHP-SAPI `apache_websocket_stream`.  The boilerplate is consistent in all the examples on the internet except for `"Content-Length: 0”` and `"Content-Type: application/octet-stream”` headers, this causes apache to disable its output buffer and output filter (compression encoding) and allows our binary data to pass.

The alternative to WebSockets is Long Polling, aka request interval polling. This is popular with CMS like WordPress due to its reliability, but is considered a poor option that only emulates real-time io with significant roundtrip overhead. 
I believe this trade-off is made all too often by developers who: 
1. Want to make sure it works everywhere
2. Can’t afford a dedicated server with `sudo` access. 
3. Have difficulties with `mod_proxy` or other Apache configurations that have to happen for a proper CLI implementation 


This method is designed to be a zero configuration open with no external Apache Modules or configurations. Using it over long term request reties will help save organizations electricity and allowing the features to be available out of the box should help shorten the learning curve for the community. If this proposal is well received then I believe the next logical step is to add the protocol framing layer in php-src too.

RichardTMiles committed:
Adding PHP_FUNCTION apache_connection_stream to expose ap_get_conn_socket

Apache CGI doesn't support WebSocket Upgrade requests out-of-the-box. php://input uses apache brigade/buckets but these are not fired with websocket Input; moreover, the php://input stream is cached and not seekable. php://stdio is the php code buffer so it is not useful here either. This allow a direct connection to the client.

RichardTMiles committed:
Adding PHP_FUNCTION apache_connection_stream to expose ap_get_conn_socket for CGI WebSockets

RichardTMiles committed:
Added the apache_connection_stream() function for CGI WebSockets

RichardTMiles committed:
Cleaning commit, removing import

RichardTMiles committed:
updating branch name

nielsdos commented:
> I’m having difficulty registering for a wiki account at https://wiki.php.net/start?do=register as I only receive
That wasn't the answer we were expecting, despite multiple browser and email options.

What email address did you fill in for the last input box? I'm guessing you perhaps filled in your own email address?

RichardTMiles commented:
@nielsdos  I've attempted my personal (Richard@Miles.Systems) and my spam (tmiles199@gmail.com). I've attempted them both at the same time too.

<img width="450" alt="Screenshot 2024-04-25 at 4 24 12 PM" src="https://github.com/php/php-src/assets/9538357/4dab52bc-01e7-4ff4-a1ab-68587a844477">

RichardTMiles committed:
untracking .DS_Store

RichardTMiles committed:
untracking .DS_Store

RichardTMiles committed:
error handling

RichardTMiles committed:
order of ZEND_PARSE_PARAMETERS_NONE

RichardTMiles committed:
PHP CLISERVER apache_connection_stream function

When a connection is received you must hit enter in your server terminal to complete the connection. CLI server is single threaded, so only basic debugging can be done regarding websockets.

vdelau commented on .gitignore:
```diff
@@ -19,6 +19,10 @@
 # Various temporary generated files
 *.tmp
 
+# editor includes
```
See the comment at the top of this file: IDE specific files should be in a local .gitignore.

vdelau commented on ext/standard/php_fopen_wrapper.c:
```diff
@@ -215,23 +215,20 @@ php_stream * php_stream_url_wrap_php(php_stream_wrapper *wrapper, const char *pa
 	if (!strcasecmp(path, "input")) {
-		php_stream_input_t *input;
-
+        php_stream_input_t *input;
```
See https://github.com/php/php-src/blob/master/CODING_STANDARDS.md#syntax-and-indentation

vdelau commented on sapi/apache2handler/php_functions_arginfo.h:
```diff
@@ -41,6 +41,7 @@ ZEND_FUNCTION(apache_lookup_uri);
 ZEND_FUNCTION(virtual);
 ZEND_FUNCTION(apache_request_headers);
 ZEND_FUNCTION(apache_response_headers);
+ZEND_FUNCTION(apache_connection_stream);
```
See comment on top of this file. This file is generated from the .stub.php file and should not be manually edited.

vdelau reviewed (commented):
I can't really comment on the technical aspect of this PR, but there are some non-technical issues I've spotted.

RichardTMiles committed:
Clean up apache_connection_stream PR

RichardTMiles committed:
Drop unrelated wrapper and gitignore changes
